### PR TITLE
COM-1787: Full pagination pattern

### PR DESCRIPTION
**Pull request by DavidHughes**
This is the new shiny Full Pagination pattern.

As documented in pagination.md, pagination.js is not a part of the pattern itself and serves to show the pagination logic that this pattern uses. This logic should be implemented in the backend to avoid rendering unnecessary markup.

### CODE REVIEW

File: assets/scss/base/_variables.scss [hunk 1]
```diff
@@ -59,6 +59,7 @@ $repo-private-icon: #e9dba5;
 
 // Blue.
 $brand-primary:         #2e008b;
+$brand-primary-gloss:   #6C5C99;
```
Comment (heylookalive) [inline assets/scss/base/_variables.scss]: Can we get this added to colours.md please!

File: docs/threads.md [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+
+Note that the pages that are hidden via inline styles should not be rendered into the pages markup if possible and are only used for the purpose of making this interactive via JavaScript. JavaScript for this makes use of the [History API](https://developer.mozilla.org/en-US/docs/Web/API/History_API) and thus does not support IE9.
+
+{% example html %}
+<nav class="thread-pagination">
+  <a href="#" class="thread-pagination__previous" aria-label="Previous page">Prev</a>
+  <li class="thread-pagination__page is-first-page"><a href="#" class="js-thread-pagination-page" data-target-page="1" aria-label="First page">1</a></li>
+  <li class="thread-pagination__page"><a href="#" class="js-thread-pagination-page" data-target-page="2"><span class="sr-only">Page </span>2</a></li>
+  <span class="thread-pagination__ellipsis--first" aria-hidden="true">...</span>
+  <li class="thread-pagination__page"><a href="#" class="js-thread-pagination-page" data-target-page="3"><span class="sr-only">Page </span>3</a></li>
```
Comment (heylookalive) [inline docs/threads.md]: Can we retitle "thread-pagination__page" to "thread-pagination__item", think that might make more sense given the it's on the li/list item.

File: assets/scss/base/_variables.scss (hunk 1 above)
Comment (DavidHughes) [inline assets/scss/base/_variables.scss]: Will do, will have a chat with Dan to get the other "glossy" variants. These are used for hover states fyi.

File: docs/threads.md (hunk 2 above)
Comment (DavidHughes) [inline docs/threads.md]: Good call.

Comment (DavidHughes) [inline docs/threads.md]: Will do this for all references to 'page' in classes, data attributes & code. I'll leave the ARIA labels as is for now until we check with Dan whether this is a thread specific pattern.

**Comment by heylookalive** [comment]
As a general note - I won't approve this pattern until there's a rationale as to why it's thread specific only.

File: assets/js/cruk-base/thread-pagination.js [hunk 3]
```diff
@@ -0,0 +1,266 @@
...
+    // Update the current page.
+    $currentPage.removeClass('thread-pagination__page--current').addClass('thread-pagination__page');
+    $('[data-target-page=' + targetPage + ']').parent('.thread-pagination__page').removeClass('thread-pagination__page').addClass('thread-pagination__page--current');
+
+    // Hide all pages and reveal the adjacent ones.
+    $('.thread-pagination__page').hide();
+
+    getAdjacentPages(targetPage).forEach(function(element) {
+      // Unless near either side of pagination, add a class to some of the outlier links.
+      // Then cry.
```
Comment (DavidHughes) [inline assets/js/cruk-base/thread-pagination.js]: Oops. Commenting here to make sure I remove this :)

**Comment by DavidHughes** [comment]
See above commits, this is now a general Pagination pattern. The part that initially made this a thread-specific thing was the concept of showing the last page button, and that button being labelled with the page number itself. This sort of pattern wouldn't be useful on things like search where later pages have less relevance.

File: assets/scss/base/_variables.scss (hunk 1 above)
Comment (DavidHughes) [inline assets/scss/base/_variables.scss]: Done!

**Comment by heylookalive** [comment]
This is much better! Please can we move the JS to /docs/assets/js/docs.js as this code is only to support the example in the documentation?

**Comment by heylookalive** [comment]
Ignore that last one, misread the path!

**Comment by DavidHughes** [comment]
Going to close this as I'll need to do some work on this to change the logic of the pagination. We've agreed that the pagination should show a maximum of 9 pages on desktop and 5 on mobile. This affects how ellipsis must be handled as they're no longer a simple Y/N.

**Pull request closed without merging**